# **Homework 2**
Name: Maddox Guthrie \
Id: 12184185

## 1.1 Exercise 1: Taylor Series Approximation
Use SymPy to compute the Taylor series of $f(x)$ about $0$ with the first $10$ non-zero
terms.

In [29]:
from sympy import symbols, exp, sin, series

x = symbols("x")

f = exp(x) * sin(2 * x)

x0 = 0
n = 11

f_series = series(f, x, x0, n)
t_series = f_series.removeO()
t_series

-779*x**10/907200 - 359*x**9/181440 + x**8/120 + 139*x**7/2520 + 11*x**6/180 - 19*x**5/60 - x**4 - x**3/3 + 2*x**2 + 2*x

Use the Taylor polynomial to approximate $f(0.4)$.

In [21]:
f_approx = t_series.subs({x:0.4})
f_approx

1.07016953578599

Compute the exact value of $f(0.4)$ using `math` package.

In [26]:
import math
f_actual = math.exp(0.4) * math.sin(2 * 0.4)
f_actual

1.0701695334073043

Compute the absolute error between the approximation and the actual value of the function
at $x = 0.4$.

In [27]:
err = abs(f_approx - f_actual)
err

2.37868547081632e-9

The approximation is very accurate. The absolute error is only $~2.38 \times 10^{-9}$, so the Taylor series matches $e^x sin(2x)$ to 8 or 9 significant figures at $x = 0.4$. This makes sense since $x = 0.4$ is close to the expansion point, so the series converges fast and the truncated terms are tiny.

## 1.2 Exercise 2: Ordinary Differential Equations

### 1.2.1 Part A: Second-Order ODE with Forcing Term
Solve the non-homogeneous equation:
$$
y'' - 3y' + 2y = \sin(2t) - t^3 + e^t
$$

In [34]:
from sympy import symbols, Function, Eq, Derivative, dsolve

t = symbols("t")
y = symbols("y", cls=Function)

eq = Eq(Derivative(y(t), t, 2) - 3 * Derivative(y(t), t, 1) + 2 * y(t), sin(2 * t) - t ** 3 + exp(t))
solns = dsolve(eq, y(t))
solns

Eq(y(t), C2*exp(2*t) - t**3/2 - 9*t**2/4 - 21*t/4 + (C1 - t)*exp(t) - sin(2*t)/20 + 3*cos(2*t)/20 - 45/8)

### 1.2.2 Part B: Second-Order ODE with Initial Conditions
Solve the equation in Part A subject to:
$$
y(0) = 0, \quad y'(0) = 1
$$

In [39]:
solns = dsolve(eq, y(t), ics={y(t).subs(t, 0): 0, y(t).diff(t).subs(t, 0): 1})
solns

Eq(y(t), -t**3/2 - 9*t**2/4 - 21*t/4 + (18/5 - t)*exp(t) + 15*exp(2*t)/8 - sin(2*t)/20 + 3*cos(2*t)/20 - 45/8)

### 1.2.3 Part C: System of First-Order ODEs (Complex Eigenvalues)
Solve the system:
$$
\begin{cases}
x'(t) = x - 2y + sin(2t) \\
y'(t) = 2x + y + t^2
\end{cases}
$$

In [42]:
x = symbols("x", cls=Function)
y = symbols("y", cls=Function)
t = symbols("t")

eq1 = Eq(x(t).diff(t), x(t) - 2 * y(t) + sin(2 * t))
eq2 = Eq(y(t).diff(t), 2 * x(t) + y(t) + t ** 2)

